### PR TITLE
Add Clamp fusion transformation

### PR DESCRIPTION
**Pull request by mateusztabaka**
It fuses Maximum->Minimum subgraph to Clamp operator.

Ticket: 44783

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ClampFusion, "ClampFusion", 0);
+
+ngraph::pass::ClampFusion::ClampFusion() {
+    auto data_pattern = ngraph::pattern::any_input();
+    auto min_const_pattern = ngraph::pattern::wrap_type<opset5::Constant>();
+    auto max_const_pattern = ngraph::pattern::wrap_type<opset5::Constant>();
+    auto max_pattern = ngraph::pattern::wrap_type<opset5::Maximum>({data_pattern, min_const_pattern});
```
Comment (mbencer) [inline inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:22]: Probably it doesn't break matching patterns, but shouldn't `max_pattern` take `max_const_pattern` as argument? (the same for `min_pattern`)

File: inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    auto min_pattern = ngraph::pattern::wrap_type<opset5::Minimum>({max_pattern, max_const_pattern});
+
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto pattern_map = m.get_pattern_value_map();
+        auto data = pattern_map.at(data_pattern);
+        auto min_const = std::dynamic_pointer_cast<opset5::Constant>(pattern_map.at(min_const_pattern).get_node_shared_ptr());
+        if (!min_const)
+            return false;
+        if (shape_size(min_const->get_shape()) != 1)
```
Comment (mbencer) [inline inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:31]: It returns with false for scalar case. You can check `is_scalar(min_const->get_shape())` (or change condition to `shape_size(min_const->get_shape()) >= 1`)

File: inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            return false;
+        auto max_const = std::dynamic_pointer_cast<opset5::Constant>(pattern_map.at(max_const_pattern).get_node_shared_ptr());
+        if (!max_const)
+            return false;
+        if (shape_size(max_const->get_shape()) != 1)
```
Comment (mbencer) [inline inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:36]: as above

**Review by mbencer — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp (hunk 1 above)
Comment (mateusztabaka) [reply inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:22]: no, that's not the way clamp behaves. Clamp does `minimum(maximum(input, min_val), max_val)`

Comment (mbencer) [reply inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:22]: ^^ right it makes sense ;)

File: inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp (hunk 2 above)
Comment (mateusztabaka) [reply inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:31]: no it doesn't. shape_size for scalars is 1 (https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/shape.hpp#L71). I could do is_scalar(...), but I wanted to allow both Shape{1} and Shape{} to match this pattern.

**Review by mbencer — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        double min_value = min_const->cast_vector<double>()[0];
+        double max_value = max_const->cast_vector<double>()[0];
+
+        auto clamp = std::make_shared<ngraph::opset5::Clamp>(data, min_value, max_value);
+        auto minimum = pattern_map.at(min_pattern);
+        clamp->set_friendly_name(minimum.get_node()->get_friendly_name());
+
+        copy_runtime_info({
+                            data.get_node_shared_ptr(),
+                            min_const,
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/clamp_fusion.cpp:48]: You don't need to copy runtime info from Constants.

File: inference-engine/tests/functional/inference_engine/transformations/clamp_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    auto res = compare_functions(f, f_ref);
+    ASSERT_TRUE(res.first) << res.second;
+}
+
+TEST(TransformationTests, ClampFusionNonConstMin) {
+    std::shared_ptr<Function> f(nullptr), f_ref(nullptr);
+    {
+        auto data = std::make_shared<opset5::Parameter>(element::f32, Shape{2, 2});
+        auto min_val = std::make_shared<opset5::Parameter>(element::f32, Shape{});
+        auto max_const = opset5::Constant::create(element::f32, Shape{}, {5});
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/clamp_fusion.cpp:58]: Please add or modify existing test to check that constants with Shape{1} also work.

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Review by postrational — APPROVED** [review]

**Review by GlebKazantaev — APPROVED** [review]

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**